libnmstate/state.py

```diff
@@ -241,6 +241,47 @@ def merge_interfaces(self, other_state):
             dict_update(other_state.interfaces[name], self.interfaces[name])
             self._ifaces_state[name] = other_state.interfaces[name]
 
+    def merge_route_config(self, other_state):
+        """
+        Merge config routes from other state to self.
+        Return a list of interface's name where route changed.
+        """
+        # Merge other_routes
+        self_iface_route_sets = defaultdict(set)
+        other_iface_route_sets = defaultdict(set)
+
+        for route in self._config_routes:
```

Comment:
If I understand correctly, this section indexes the routes and keeps values as RouteEntry objects.
So without understanding the whole picture here (seems to me like storing so many duplications (3) of the same data is a smell), you could just create a method that goes over the indexed data (which is available already) and replaces the dicts with the objects.

BTW, you could consider here to actually index the objects instead of the dict (in the constructor) and only convert it to a dict if you must (like in the getter upon request). The object itself could go down the stack, it will be better than using dicts. Just consider this.

Reply:
The `metadata.py` will use the `state.config_iface_routes` where I don't want expose RouteEntry to. So dict it will be.